marlowe-dashboard-client/src/Capability/Marlowe.purs

```diff
@@ -102,6 +107,32 @@ instance manageMarloweAppM :: ManageMarlowe AppM where
             , previousCompanionAppState: Nothing
             }
         pure $ createWalletDetails <$> ajaxCompanionAppId <*> ajaxMarloweAppId <*> ajaxAssets
+  restoreWallet options = do
+    mWalletInfo <- Wallet.restoreWallet options
+    case mWalletInfo of
+      Left err -> pure $ Left err
+      Right walletInfo -> do
+        let
+          walletId = view _walletId walletInfo
```

Comment:
I'm not sure if I understand the context and the comment:

`Variant` based error handling allow us to compose errors from different independent parts of the application into a single type (no need for nesting errors).
Adding more more detailed information / semantics to errors is a bit orthogonal problem I think and can be done without _checked-exceptions_.

Reply:
Well, handling errors is tricky. 
For an end user, if we cannot restore a wallet, of the 3 possible error options that I modelled, only one gives the user something meaningful to do:

```purescript
data RestoreError
  = InvalidMnemonic
  | ClientServerError AjaxError
  | ServerError BE.RestoreError
```
If the user provides an invalid mnemonic phrase because he/she made out words, then we provide a message saying that the phrase is invalid, and the user can try again with a new phrase.

The other two is either a Client to Server ajax error that we are not catching (any non 2xx message), or the server failing somehow (maybe the wallet backend is not available). 

In any case, it's probably not worth it to provide a message more elaborated than "There was a problem, try again later or contact support". 

If that is the approach we want to take for Client/Server errors, then this could be a Variant of 3 individual errors. Invalid mnemonic would only apply to restore wallet, and the other two could be re-utilized for other endpoints (for example the getTotalFunds that I'm working on). 

Showing a message to a user is the bare minimum we can do to avoid showing a loading icon forever, but it doesn't help us diagnose or fix the error. That is where observability comes into play, and tools like getsentry are super useful. But we need to add/aggregate enough information to be able to work offline. Some of the context is manually added by the tools (browser information, list of pages visited) and we could add our own information, for example a list of Actions that were executed.
In this case in particular, we are treating Client/Server communication problems the same, and we are not even saying which step did it fail, was it on the restoreWallet endpoint? Was it when trying to activate the wallet companion? when activating the MarloweApp?

We could probably create a variant that is ClientServer communication, in which we allow to add a name of step, arbitrary JSON context and the Ajax error itself.

In general, I was imagining on making the `handleAction` to work with a `MonadError` of a Variant in which all members need to implement a `ToSentry` typeclass (not sure if it's possible to model that), and maybe a `ToToast` typeclass. That way, we can create some variants like InvalidMnemonic that needs to be cached, and for the rest, we can provide a general handler that sends the error to sentry and puts a toast message